### DISCUSSION TITLE
Image Upload error in docker/kubernetes

### DISCUSSION THREAD

**Discussion by Judhianto** [serenity-is/Serenity#5373] [Help]
I am using serenity.net 5 with visualstudio 2019

I implement ImageUploadEditor on my xxx_row.cs definition . 

When running in iisexpress for testing, there is no problem with upload image process, i can upload image and can displayed later in the record view. 

But when i build as docker image and run on my kubernetes server, upload image failed without any notification, i also try run on docker in my pc with the same problem. Anything else just works fine.

I suspect there was some problem with path configuration, since in the test i am using iisexpress on windows, but in deployment i am using kubernetes on linux. maybe different style in path definition between windows and linux, create the problem.

For uploadpath, i didn't make any changes in file appsettings.json
/////
  "UploadSettings": {
    "Path": "~/App_Data/upload/"
  },
/////

So, what configuration must be changed in kubernetes/docker deployment?

**Comment by VictorTomaili** [comment]
When you run a project in container, you will want to use persistent storage. Write your persistent storage path in config will fix that. If you still get error you may have some docker configuration mistakes as well. Maybe you need to write your own file upload services. Check `DefaultUploadStorage` file for a sample. Then you can upload your file to external storage temporary & persistent like azure blob storage etc.

**Reply by Judhianto** [reply]
I already did to use persistent storage, and while run shell in the container, i can write-edit-delete file in the persistent storage.

below the .yaml configuration for load the container to my kubernetes server


---
apiVersion: v1
kind: PersistentVolume
metadata:
  name: pv-eosweb
spec:
  capacity:
    storage: 1.1Gi
  accessModes:
    - ReadWriteMany
  nfs:
    server: 10.152.183.40
    path: "/eos/web"

---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: pvc-eosweb
spec:
  storageClassName: ""
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 1.1Gi

---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: eosweb
  labels:
    app: web-eosweb
spec:
  replicas: 2
  selector:
    matchLabels:
      app: web-eosweb
  strategy:
    type: RollingUpdate
    rollingUpdate:
      maxSurge: 1
      maxUnavailable: 0
  template:
    metadata:
      labels:
        app: web-eosweb
    spec:
      containers:
      - name: eoswebcontainer
        image: judhianto/eosweb:1.5
        imagePullPolicy: "Always"
        ports:
        - containerPort: 80
        volumeMounts:
        - name: storage-eosweb
          mountPath: /App_Data
      volumes:
      - name: storage-eosweb
        persistentVolumeClaim:
          claimName: pvc-eosweb

---
kind: Service
apiVersion: v1
metadata:
  name: web-eosweb
  labels:
    app: web-eosweb
spec:
  selector:
    app: web-eosweb
  ports:
  - protocol: TCP
    port: 80
    targetPort: 80
  clusterIP: None

**Reply by Judhianto** [reply]
Another failed try.

I created new project with serenity template. I install using default options, I didn't change any code.

I add docker support, then build the solution.

When I run in docker or kubernetes, I have the same problem >> can not upload image file.

For testing, i upload picture into Product record in Northwind database. it failed.

So, any idea to fix this image upload problem?

**Reply by VictorTomaili** [reply]
Try to check installed libraries in your docker container. 
https://github.com/dotnet/core/issues/2746#issuecomment-595980412

**Reply by Judhianto** [reply]
> Try to check installed libraries in your docker container.
> [dotnet/core#2746 (comment)](https://github.com/dotnet/core/issues/2746#issuecomment-595980412)

Thank you! it's solve the problem

Below is my final dockerfile

FROM mcr.microsoft.com/dotnet/aspnet:5.0-buster-slim AS base
RUN apt-get update -yq \
    && apt-get install curl gnupg -yq \
    && curl -sL https://deb.nodesource.com/setup_15.x | bash \
    && apt-get install nodejs -yq \
    && apt-get install apt-utils libgdiplus libc6-dev -yq

WORKDIR /app
EXPOSE 80

FROM mcr.microsoft.com/dotnet/sdk:5.0-buster-slim AS build
RUN apt-get update -yq \
    && apt-get install curl gnupg -yq \
    && curl -sL https://deb.nodesource.com/setup_15.x | bash \
    && apt-get install nodejs -yq\
    && apt-get install apt-utils libgdiplus libc6-dev -yq

WORKDIR /src
COPY ["EOS.Web/EOS.Web.csproj", "EOS.Web/"]
RUN dotnet restore "EOS.Web/EOS.Web.csproj"
COPY . .
WORKDIR "/src/EOS.Web"
RUN dotnet build "EOS.Web.csproj" -c Release -o /app/build

FROM build AS publish
RUN dotnet publish "EOS.Web.csproj" -c Release -o /app/publish

FROM base AS final
WORKDIR /app
COPY --from=publish /app/publish .
ENTRYPOINT ["dotnet", "EOS.Web.dll"]

**Reply by edwardch** [reply]
Hey @Judhianto, I see you have your `base` and `build` images doing the same things.

I was able to get away with the following:

```bash
# .NET Core 3.1 version of Serenity & Serene
FROM mcr.microsoft.com/dotnet/core/aspnet:3.1-buster-slim AS base
# install System.Drawing native dependencies
RUN apt-get update \
    && apt-get install -y --allow-unauthenticated \
        libc6-dev \
        libgdiplus \
        libx11-dev \
     && rm -rf /var/lib/apt/lists/*

WORKDIR /app
EXPOSE 80

FROM mcr.microsoft.com/dotnet/core/sdk:3.1-buster AS build
RUN apt-get update -yq && apt-get upgrade -yq && apt-get install -yq curl git nano
RUN curl -sL https://deb.nodesource.com/setup_10.x | bash - && apt-get install -yq nodejs 
RUN npm install -g npm

WORKDIR /src
# ...
```
I probably need to tidy a couple of things up and I haven't yet tested with v5 and 5.0-buster-slim, but you might get better build speeds by removing the extra stuff. 

Let me know if it helps, and let me know if I've done something glaringly stupid.

Cheers,
Edward

**Comment by alaamh** [comment]
it works for me after mapping App_Data in docker-compose
```
volumes:
      - ./volumes/app_data:/App_Data
```
and change appsettings.json
```
  "UploadSettings": {
    "Path": "/App_Data/upload/"
  }
```

Please note I have removed the `~`
